akka-stream-tests/src/test/scala/akka/stream/scaladsl/FoldWithinSpec.scala

```diff
@@ -0,0 +1,244 @@
+    timePasses(maxDuration/4)
+    p.sendNext(3)
+    timePasses(maxDuration/4)
+    p.sendNext(4)
+    timePasses(maxDuration/4) // maxDuration will accumulate
+
+    p.sendNext(5)
+    p.sendNext(6)
+    p.sendNext(7)
+    p.sendComplete()
+
+    Await.result(result, 10.seconds) should be (Seq(Seq(1, 2, 3, 4), Seq(5, 6, 7)))
+
+  }
+
+  "down stream back pressure should not miss data on completion without pull on start" in {
+
+    val maxGap = 1.second
+    val upstream = TestPublisher.probe[Int]()
+    val downstream = TestSubscriber.probe[Seq[Int]]()
+    // Note that the cost function set to zero here means the stream will accumulate elements until completed
+    Source.fromPublisher(upstream).via(
+      new FoldWithin[Int, Seq[Int], Seq[Int]](
+        seed = i => Seq(i),
+        aggregate = (seq, i) => seq :+ i,
+        emitOnAgg = _ => false,
+        harvest = seq => seq,
+        maxGap = Some(maxGap), // elements with longer gap will put put to next aggregator
+        getSystemTimeMs = getSystemTimeMs,
+        pullOnStart = false
+      )
+    ).to(Sink.fromSubscriber(downstream)).run()
+
+    downstream.ensureSubscription()
+    upstream.sendNext(1)
+    timePasses(maxGap) // this will not cause emit, without eager pull the data does not make to the next stage until downstream request
+    upstream.sendNext(2)
+    downstream.request(1)
+    downstream.expectNoMessage() // no emit
+    timePasses(maxGap) // this will cause emit
+    downstream.expectNext(Seq(1, 2))
+    upstream.sendNext(3)
+    upstream.sendNext(4)
+    downstream.request(1)
+    downstream.expectNoMessage()
+    upstream.sendComplete() // flush the pending data
+    downstream.expectNext(Seq(3, 4))
+    downstream.expectComplete()
+  }
+
+  "down stream back pressure should not miss data on completion with pull on start" in {
+
+    val maxGap = 1.second
+    val upstream = TestPublisher.probe[Int]()
+    val downstream = TestSubscriber.probe[Seq[Int]]()
+    // Note that the cost function set to zero here means the stream will accumulate elements until completed
+    Source.fromPublisher(upstream).via(
+      new FoldWithin[Int, Seq[Int], Seq[Int]](
+        seed = i => Seq(i),
+        aggregate = (seq, i) => seq :+ i,
+        emitOnAgg = _ => false,
+        harvest = seq => seq,
+        maxGap = Some(maxGap), // elements with longer gap will put put to next aggregator
+        getSystemTimeMs = getSystemTimeMs,
+        pullOnStart = true
+      )
+    ).to(Sink.fromSubscriber(downstream)).run()
+
+    downstream.ensureSubscription()
+    upstream.sendNext(1)
+    timePasses(maxGap) // with pull on start this will cause emit later
+    upstream.sendNext(2) // this will be back pressured
+    timePasses(maxGap) // this gap will not cause emit due to back pressure
+    downstream.request(1)
+    downstream.expectNext(Seq(1)) // emit due to the first gap
+    upstream.sendNext(3)
+    downstream.request(1)
+    downstream.expectNoMessage()
```

Comment:
I wonder if `expectNoMessage()` calls shouldn't use an empty duration rather than the default since we are using simulated elapsed time anyway

Reply:
without time parameter it's just using the default time `testKitSettings.ExpectNoMessageDefaultTimeout.dilated`
time doesn't matter in this test case.
any concerns here?